local_planner/src/nodes/planner_functions.cpp

```diff
@@ -308,13 +308,56 @@ void getCostMatrix(const Histogram& histogram, const Eigen::Vector3f& goal,
     }
   }
 
-  float smoothing_margin_degrees = 45;
+  float smoothing_margin_degrees = 30;
   unsigned int smooth_radius = ceil(smoothing_margin_degrees / ALPHA_RES);
   smoothPolarMatrix(distance_matrix, smooth_radius);
 
+  generateCostImage(cost_matrix, distance_matrix, image);
   cost_matrix = cost_matrix + distance_matrix;
 }
 
+void generateCostImage(const Eigen::MatrixXf& cost_matrix,
+                       const Eigen::MatrixXf& distance_matrix,
+                       sensor_msgs::Image& image) {
+  float max_val = std::max(cost_matrix.maxCoeff(), distance_matrix.maxCoeff());
+  image.height = GRID_LENGTH_E;
+  image.width = GRID_LENGTH_Z;
+  image.encoding = "rgb8";
+  image.is_bigendian = 0;
+  image.step = 3 * image.width;
+
+  int image_size = static_cast<int>(image.height * image.width);
+  image.data.clear();
+  image.data.reserve(3 * image_size);
+
+  for (int e = GRID_LENGTH_E - 1; e >= 0; e--) {
+    for (int z = 0; z < GRID_LENGTH_Z; z++) {
+      float distance_cost = 255.f * distance_matrix(e, z) / max_val;
+      float other_cost = 255.f * cost_matrix(e, z) / max_val;
+      image.data.push_back((int)std::max(0.0f, std::min(255.f, distance_cost)));
+
+      image.data.push_back((int)std::max(0.0f, std::min(255.f, other_cost)));
+
+      image.data.push_back(0);
+    }
+  }
+}
+
+int colorImageIndex(int e_ind, int z_ind, int color) {
+  // color = 0 (red), color = 1 (green), color = 2 (blue)
+  int index = 0;
```

Comment:
how about `return (e_ind * GRID_LENGTH_E + z_ind)*3 + color)`

Reply:
It is actually more complicated since e_ind is reversed, so it would be something like

`return ((GRID_LENGTH_E - e_ind * GRID_LENGTH_E) + z_ind) * 3 + color;` , with perhaps an off-by-1 error in there somewhere.

Reply:
and the right answer iiiiiiiis:   ((GRID_LENGTH_E - e_ind - 1) * GRID_LENGTH_Z + z_ind) * 3 + color;